### PR TITLE
Django3: add new `django.db.models.JSONField`

### PR DESCRIPTION
**Pull request by humitos**
Create new JSON fields postfixed with `_json` for all the JSON fields defined
via `jsonfield` third party package.

```
In [2]: Build.objects.filter(_config_json__build__image='readthedocs/build:latest')
Out[2]: <BuildQuerySet [<Build: Build qtile for admin (2)>, <Build: Build qtile for admin (1)>]>

In [3]: Build.objects.filter(_config_json__build__image='readthedocs/build:latest').first()._config
Out[3]: 
{'version': '1',
 'formats': ['htmlzip', 'epub', 'pdf'],
 'python': {'version': '3',
  'install': [{'requirements': None}],
  'use_system_site_packages': False},
 'conda': None,
 'build': {'image': 'readthedocs/build:latest', 'apt_packages': []},
 'doctype': 'sphinx',
 'sphinx': {'builder': 'sphinx',
  'configuration': '/usr/src/app/checkouts/readthedocs.org/user_builds/qtile/checkouts/stable/docs/conf.py',
  'fail_on_warning': False},
 'mkdocs': {'configuration': None, 'fail_on_warning': False},
 'submodules': {'include': 'all', 'exclude': [], 'recursive': True},
 'search': {'ranking': {}, 'ignore': []}}

In [4]: Build.objects.filter(_config_json__build__image='readthedocs/build:latest').first()._config_json
Out[4]: 
{'build': {'image': 'readthedocs/build:latest', 'apt_packages': []},
 'conda': None,
 'mkdocs': {'configuration': None, 'fail_on_warning': False},
 'python': {'install': [{'requirements': None}],
  'version': '3',
  'use_system_site_packages': False},
 'search': {'ignore': [], 'ranking': {}},
 'sphinx': {'builder': 'sphinx',
  'configuration': '/usr/src/app/checkouts/readthedocs.org/user_builds/qtile/checkouts/stable/docs/conf.py',
  'fail_on_warning': False},
 'doctype': 'sphinx',
 'formats': ['htmlzip', 'epub', 'pdf'],
 'version': '1',
 'submodules': {'exclude': [], 'include': 'all', 'recursive': True}}

In [5]:
```

### CODE REVIEW

File: readthedocs/builds/models.py [hunk 1]
```diff
@@ -648,6 +648,7 @@ class Build(models.Model):
         blank=True,
     )
     _config = JSONField(_('Configuration used in the build'), default=dict)
+    _config_json = models.JSONField(_('Configuration used in the build'), default=dict)
```
Comment (ericholscher) [inline readthedocs/builds/models.py:651]: Seems like this needs a comment explaining why there's 2.

File: readthedocs/builds/models.py [hunk 2]
```diff
@@ -716,6 +717,10 @@ def config(self):
         Build object (it could be stored in this object or one of the previous
         ones).
         """
+        # TODO: now that we are using a proper JSONField here, we could
+        # probably change this field to be a ForeignKey to avoid repeating the
+        # config file over and over again and re-use them to save db data as
+        # well
```
Comment (ericholscher) [inline readthedocs/builds/models.py:723]: Yea, an FK seems reasonable but would be another field right? I agree that our current implementation is just a hacky FK :)

**Review by ericholscher — APPROVED** [review]
This looks reasonable. Is the plan to then remove the old fields, once we've converted the data? The migration path wasn't clear to me from the description or code.

File: readthedocs/builds/models.py (hunk 2 above)
Comment (humitos) [reply readthedocs/builds/models.py:723]: Yeah. I'm thinking about having something like

```
# builds.models.Build
config = ForeignKey(BuildConfig)

# config.models.BuildConfig
data = JSONField(default=None)
```

... and later in the code, I'm expecting to do:

```
build.config = BuildConfig.objects.get_or_create(data=user_config)
```

and relying on Postgres to do the work about deciding if it's the same or a new one properly  😄

**Comment by stsewd** [comment]
> If we do both at the same deploy, we won't need to resync. However, if we only do the first one, we will do to the copy for new Build objects.

This looks like it will still result in some data loss.

- Migrations adding the new field and the data migration are run
- In the meantime, more objects are created (and will fill only the old json fields, leaving the new json fields empty)
- Migration removing the old fields are run
- Objects that were created in the meantime will result with empty json fields

I think we should split the migration in two: one creating the fields, and another doing the data migration, and to avoid data loss, we should override the save() methods of the models to populate the new json fields with the information from the old json fields when a model is saved. Then the steps will be as follows:

- Run the migration that creates the new fields before deploying the webs
- Deploy the code (now new objects will be saving the same data in both fields)
- Run the data migration for the old models
- Run the migration that renames the field (I think there will be some downtime, since the original fields are being deleted before the other one is renamed)
- Since we were only setting the content of the old json fields to the new ones (we aren't accessing the field, just setting an attribute), the code on the `save()` method can be safely deleted in another deploy.

**Comment by humitos** [comment]
Good description of the potential problem :+1: 

@stsewd I updated this PR to split the migration and overwrite the `.save` method. Please, let me know if I understood your comment correctly or there is something else that's missing.

**Pull request merged**